File: src/main/java/org/eln/eln3/sim/ThermalResistor.java
package org.eln.eln3.sim;

public class ThermalResistor implements IProcess {

    ThermalLoad a;
    ThermalLoad b;

    protected double thermalResistance;
    protected double thermalResistanceInverse;

    public ThermalResistor(ThermalLoad a, ThermalLoad b) {
        this.a = a;
        this.b = b;
        highImpedance();
    }

    @Override
    public void process(double time) {
        double power = (a.temperatureCelsius - b.temperatureCelsius) * thermalResistanceInverse;
        a.netThermalPowerAccumulator -= power;
        b.netThermalPowerAccumulator += power;
    }

    public double getPower() {
        return (a.temperatureCelsius - b.temperatureCelsius) * thermalResistanceInverse;
    }

    public void setThermalResistance(double thermalResistance) {
        this.thermalResistance = thermalResistance;
        thermalResistanceInverse = 1 / thermalResistance;
    }

    public double getThermalResistance() {
        return thermalResistance;
    }

    public void highImpedance() {
        setThermalResistance(ThermalLoad.HIGH_THERMAL_RESISTANCE);
    }
}

File: src/main/kotlin/org/eln/eln3/sim/thermal/ThermalResistor.kt
package org.eln.eln3.sim.thermal

import org.eln.eln3.sim.IProcess

class ThermalResistor(var a: ThermalLoad, var b: ThermalLoad) : IProcess {
    var thermalResistance: Double = 0.0
    val thermalResistanceInverse: Double
        get() = 1 / thermalResistance

    init {
        highImpedance()
    }

    override fun process(time: Double) {
        val power = (a.temperatureCelsius - b.temperatureCelsius) * thermalResistanceInverse
        a.netThermalPowerAccumulator = a.netThermalPowerAccumulator - power
        b.netThermalPowerAccumulator = b.netThermalPowerAccumulator + power
    }

    val power: Double
        get() = (a.temperatureCelsius - b.temperatureCelsius) * thermalResistanceInverse

    fun highImpedance() {
        thermalResistance = ThermalLoad.HIGH_THERMAL_RESISTANCE
    }
}